src/hooks/swap/useApproval.ts

```diff
@@ -0,0 +1,103 @@
+import { Percent, Token } from '@uniswap/sdk-core'
+import { useWeb3React } from '@web3-react/core'
+import { SWAP_ROUTER_ADDRESSES } from 'constants/addresses'
+import { ErrorCode } from 'constants/eip1193'
+import { PermitState, SignatureData, useERC20PermitFromTrade } from 'hooks/useERC20Permit'
+import useTransactionDeadline from 'hooks/useTransactionDeadline'
+import { useCallback, useMemo } from 'react'
+import { InterfaceTrade } from 'state/routing/types'
+import { isExactInput } from 'utils/tradeType'
+
+import { AllowanceState, useAllowanceCallback, useAllowanceState } from '../useAllowance'
+
+export enum ApprovalState {
+  REQUIRES_ALLOWANCE,
+  PENDING_ALLOWANCE,
+  REQUIRES_SIGNATURE,
+  PENDING_SIGNATURE,
+  APPROVED,
+}
+
+export interface Approval {
+  state: ApprovalState
+  allowance: AllowanceState
+  signatureData?: SignatureData
+  gatherPermitSignature?: () => Promise<void>
+}
+
+function useSwapSpender() {
+  const { chainId } = useWeb3React()
+  return chainId ? SWAP_ROUTER_ADDRESSES[chainId] : undefined
+}
+
+function useAmountToAllow(trade: InterfaceTrade | undefined, allowedSlippage: Percent) {
+  return useMemo(() => {
+    if (!trade) return undefined
+    if (isExactInput(trade.tradeType)) return trade.inputAmount
+    if (!trade.inputAmount.currency.isToken) return undefined
+    return trade.maximumAmountIn(allowedSlippage)
+  }, [allowedSlippage, trade])
+}
+
+export function useApproval(
+  trade: InterfaceTrade | undefined,
+  allowedSlippage: Percent,
+  useIsPendingApproval: (token?: Token, spender?: string) => boolean
+): Approval {
+  const deadline = useTransactionDeadline()
+
+  // Check allowance on ERC20 contract based on amount.
+  const allowance = useAllowanceState(useAmountToAllow(trade, allowedSlippage), useSwapSpender(), useIsPendingApproval)
+
+  // Check permit if token supports it.
+  const {
+    state: permitState,
+    signatureData,
+    gatherPermitSignature,
+  } = useERC20PermitFromTrade(trade, allowedSlippage, deadline)
+
+  const state = useMemo(() => {
+    if (allowance === AllowanceState.PENDING) {
+      return ApprovalState.PENDING_ALLOWANCE
+    } else if (permitState === PermitState.LOADING) {
+      return ApprovalState.PENDING_SIGNATURE
+    } else if (allowance !== AllowanceState.NOT_ALLOWED || permitState === PermitState.SIGNED) {
+      return ApprovalState.APPROVED
+    } else if (gatherPermitSignature) {
+      return ApprovalState.REQUIRES_SIGNATURE
+    } else {
+      return ApprovalState.REQUIRES_ALLOWANCE
+    }
```

Comment:
nit: (might not even be worth addressing) it's possible that a user has given sufficient approval to a token that implements permit

Reply:
(so requesting a permit signature wouldn't be necessary)

Reply:
this should be considered APPROVED using the current PR logic